cpp/oneapi/dal/backend/primitives/rng/dpc_engine.hpp

```diff
@@ -0,0 +1,233 @@
+* limitations under the License.
+*******************************************************************************/
+
+#pragma once
+
+#include "oneapi/dal/backend/primitives/rng/utils.hpp"
+#include "oneapi/dal/backend/primitives/rng/rng_types.hpp"
+
+#include <daal/include/algorithms/engines/mt2203/mt2203.h>
+#include <daal/include/algorithms/engines/mcg59/mcg59.h>
+#include <daal/include/algorithms/engines/mrg32k3a/mrg32k3a.h>
+#include <daal/include/algorithms/engines/philox4x32x10/philox4x32x10.h>
+#include <daal/include/algorithms/engines/mt19937/mt19937.h>
+
+#include <oneapi/mkl.hpp>
+
+namespace mkl = oneapi::mkl;
+namespace oneapi::dal::backend::primitives {
+
+#ifdef ONEDAL_DATA_PARALLEL
+
+template <engine_method EngineType>
+struct dpc_engine_type;
+
+template <>
+struct dpc_engine_type<engine_method::mt2203> {
+    using type = oneapi::mkl::rng::mt2203;
+};
+
+template <>
+struct dpc_engine_type<engine_method::mcg59> {
+    using type = oneapi::mkl::rng::mcg59;
+};
+
+template <>
+struct dpc_engine_type<engine_method::mt19937> {
+    using type = oneapi::mkl::rng::mt19937;
+};
+
+template <>
+struct dpc_engine_type<engine_method::mrg32k3a> {
+    using type = oneapi::mkl::rng::mrg32k3a;
+};
+
+template <>
+struct dpc_engine_type<engine_method::philox4x32x10> {
+    using type = oneapi::mkl::rng::philox4x32x10;
+};
+
+/// A class that provides a unified interface for random number generation on both CPU and GPU devices.
+///
+/// This class serves as a wrapper for random number generators (RNGs) that supports different engine types,
+/// enabling efficient random number generation on heterogeneous platforms using SYCL. It integrates a host
+/// (CPU) engine and a device (GPU) engine, allowing operations to be executed seamlessly on the appropriate
+/// device.
+///
+/// @tparam EngineType The RNG engine type to be used. Defaults to `engine_method::mt2203`.
+///
+/// @param[in] queue The SYCL queue used to manage device operations.
+/// @param[in] seed  The initial seed for the random number generator. Defaults to `777`.
+///
+/// The class provides functionality to skip ahead in the RNG sequence, retrieve engine states, and
+/// manage host and device engines independently. Support for `skip_ahead` on GPU is currently limited for
+/// some engine types.
+template <engine_method EngineType = engine_method::mt2203>
+class dpc_engine {
+public:
+    using dpc_engine_t = typename dpc_engine_type<EngineType>::type;
+
+    explicit dpc_engine(sycl::queue& queue, std::int64_t seed = 777)
+            : q(queue),
+              host_engine_(initialize_host_engine(seed)),
+              dpc_engine_(initialize_dpc_engine(queue, seed)),
+              impl_(dynamic_cast<daal::algorithms::engines::internal::BatchBaseImpl*>(
+                  host_engine_.get())) {
+        if (!impl_) {
+            throw std::domain_error("RNG engine is not supported");
+        }
+    }
+
+    virtual ~dpc_engine() = default;
```

Comment:
This class violates the rule of three/five. Because it implements the destructor, but does not implement copy/move constructor and assignment/move operator.
Please specify those operators as `=default` or `=delete`.
https://scottmeyers.blogspot.com/2014/03/a-concern-about-rule-of-zero.html